fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java

```diff
@@ -653,70 +660,149 @@ public static List<Range<PartitionKey>> mergeRanges(List<Range<PartitionKey>> ra
+            ScalarOperator partitionPredicate = Utils.compoundOr(rangePredicates);
+            if (partitionPredicate != null) {
+                partitionPredicates.add(partitionPredicate);
+            }
+        } else {
+            return null;
         }
+
+        return partitionPredicates;
+    }
+
+
+    public static List<ScalarOperator> compensatePartitionPredicateForHiveScan(LogicalHiveScanOperator scanOperator,
+                                                                               ColumnRefFactory columnRefFactory) {
         List<ScalarOperator> partitionPredicates = Lists.newArrayList();
-        for (LogicalOlapScanOperator olapScanOperator : olapScanOperators) {
-            Preconditions.checkState(olapScanOperator.getTable().isNativeTable());
-            OlapTable olapTable = (OlapTable) olapScanOperator.getTable();
+        Preconditions.checkState(scanOperator.getTable().isHiveTable());
+        HiveTable hiveTable = (HiveTable) scanOperator.getTable();
 
-            if (olapTable.getPartitionInfo() instanceof SinglePartitionInfo) {
-                continue;
+        if (hiveTable.isUnPartitioned()) {
+            return partitionPredicates;
+        }
+
+        ScanOperatorPredicates scanOperatorPredicates = scanOperator.getScanOperatorPredicates();
+        if (scanOperatorPredicates.getSelectedPartitionIds().size() ==
+                scanOperatorPredicates.getIdToPartitionKey().size()) {
+            return partitionPredicates;
+        }
+
+        if (scanOperatorPredicates.getSelectedPartitionKeys().stream().anyMatch(partitionKey -> {
+            // only support one partition column now.
+            if (partitionKey.getKeys().size() != 1) {
+                return true;
             }
+            LiteralExpr e = partitionKey.getKeys().get(0);
+            // Only support date/int type
+            if (!(e instanceof DateLiteral || e instanceof IntLiteral)) {
+                return true;
+            }
+            return false;
+        })) {
+            return partitionPredicates;
+        }
 
-            if (olapScanOperator.getSelectedPartitionId() != null
-                    && olapScanOperator.getSelectedPartitionId().size() == olapTable.getPartitions().size()) {
-                continue;
+        List<Range<PartitionKey>> ranges = Lists.newArrayList();
+        for (PartitionKey selectedPartitionKey : scanOperatorPredicates.getSelectedPartitionKeys()) {
+            try {
+                LiteralExpr expr = PartitionUtil.addOffsetForLiteral(selectedPartitionKey.getKeys().get(0), 1);
+                PartitionKey partitionKey = new PartitionKey(ImmutableList.of(expr), selectedPartitionKey.getTypes());
+                ranges.add(Range.closedOpen(selectedPartitionKey, partitionKey));
+            } catch (AnalysisException e) {
+                LOG.warn("Compute partition key range failed. ", e);
```

Comment:
why return empty list when exception? why not throw the exception?

Reply:
Because it doesn't matter the query rewrite, the partition predicates are not erased by the rule for hive table